server/lib/src/idm/ldap.rs

```diff
@@ -823,6 +865,591 @@ mod tests {
+            );
+
+            let ct = duration_from_epoch_now();
+            let mut server_txn = idms.proxy_write(ct).await;
+            assert!(server_txn
+                .qs_write
+                .internal_create(vec![e1, e2, e3])
+                .and_then(|_| server_txn.commit())
+                .is_ok());
+        }
+
+        // No session, app does not have linked group
+        let res = ldaps
+            .do_bind(idms, "spn=testperson1,app=testapp1,dc=example,dc=com", "")
+            .await;
+        assert!(res.is_ok());
+        assert!(res.unwrap().is_none());
+
+        {
+            let ml = ModifyList::new_purge_and_set(Attribute::LinkedGroup, Value::Refer(grp_uuid));
+            let mut idms_prox_write = idms.proxy_write(duration_from_epoch_now()).await;
+            assert!(idms_prox_write
+                .qs_write
+                .internal_modify_uuid(app_uuid, &ml)
+                .is_ok());
+            assert!(idms_prox_write.commit().is_ok());
+        }
+
+        // No session, user not member of linked group
+        let res = ldaps
+            .do_bind(idms, "spn=testperson1,app=testapp1,dc=example,dc=com", "")
+            .await;
+        assert!(res.is_ok());
+        assert!(res.unwrap().is_none());
+
+        {
+            let ml = ModifyList::new_append(Attribute::Member, Value::Refer(usr_uuid));
+            let mut idms_prox_write = idms.proxy_write(duration_from_epoch_now()).await;
+            assert!(idms_prox_write
+                .qs_write
+                .internal_modify_uuid(grp_uuid, &ml)
+                .is_ok());
+            assert!(idms_prox_write.commit().is_ok());
+        }
+
+        // No session, user does not have app password for testapp1
+        let res = ldaps
+            .do_bind(idms, "spn=testperson1,app=testapp1,dc=example,dc=com", "")
+            .await;
+        assert!(res.is_ok());
+        assert!(res.unwrap().is_none());
+
+        let pass1: String;
+        let pass2: String;
+        let pass3: String;
+        {
+            let mut idms_prox_write = idms.proxy_write(duration_from_epoch_now()).await;
+
+            let ev = GenerateApplicationPasswordEvent::new_internal(
+                usr_uuid,
+                app_uuid,
+                "apppwd1".to_string(),
+            );
+            pass1 = idms_prox_write
+                .generate_application_password(&ev)
+                .expect("Failed to generate application password");
+
+            let ev = GenerateApplicationPasswordEvent::new_internal(
+                usr_uuid,
+                app_uuid,
+                "apppwd2".to_string(),
+            );
+            pass2 = idms_prox_write
+                .generate_application_password(&ev)
+                .expect("Failed to generate application password");
+
+            assert!(idms_prox_write.commit().is_ok());
+
+            // event silently dropped on duplicated label (check Entry::add_ava())
```

Comment:
Rather than silent drop, I think it should either error or overwrite. Silently losing a change is user hostile.

Reply:
I changed the code to overwrite on duplicated label for now.